### PR TITLE
Convert dashboard to use esbuild

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Converts `dashboard/` module to use esbuild instead of create-react-app + webpack
- Removes some unused dependencies
- Upgrades esbuild from `^0.15.14` to `^0.17.0` (required for esbuild's serve mode)
- Removes `enso-gui-server` (`server/`) module
- Switches from `ts-node` (`tsc`/`swc`-based) to `tsx` (`esbuild`-based) since we're switching everything else to `esbuild`
- Renames `npm run dev` to `npm run watch-dashboard`

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by somebody1234** [comment]
some notes:
1. this is close to the bare minimum required to make things work.
  a. there was (and still is, to keep changes simple) a lot of duplication in the esbuild scripts. i'd be happy to change it though
2. i've tested watch mode for `dashboard/`, `content/` (both `npm run watch`) and `client/` (`./run ide watch` and `./run ide build`), however more extensive testing is a *very* good idea since this PR changes almost all of the esbuild files
3. this is a pretty big change so not sure whether a changelog will be needed, so not adding [CI: No changelog needed] for now
4. `tailwind.css` is still generated once only, just like in the previous PR. it may be possible to get it to only build when the file on disk is changed, but may be more trouble than it's worth
  a. this should be *mostly* fine though. firstly, `tailwind.css` shouldn't change very often, and secondly, because it has been switched to `esbuild`, restarting a build from scratch shouldn't slow down the dev cycle, even on slower machines

**Comment by somebody1234** [comment]
whoops, develop was behind locally

**Comment by somebody1234** [comment]
as per latest commit message, i've renamed `npm run dev` to `npm run watch-dashboard` because i think `dev` is a little non-descriptive - especially since it isn't even clear which module is being run

File: app/ide-desktop/lib/content/start.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/start.ts]: some really strange indentation here - looks like tabs are used.

File: app/ide-desktop/lib/dashboard/src/tailwind.css
Comment (wdanilo) [inline app/ide-desktop/lib/dashboard/src/tailwind.css]: this looks like a generated file, it should not be committed.

**Review by wdanilo — COMMENTED** [review]
Oh wow, I was not expecting this to be such a big change (so many files). It makes sense after seeing the code, but somehow in my brain imagined that as a way smaller change earlier. 

Anyway, answering your points from above (I edited your comment and replaced bullet points with enumeration, so I can refer to it easily):

1. Perfect. If you will find any places that can be easily improved to minimize the code duplication, go for it, but this is not super important right now (this kind of tech debt is way less important than wrong project structure / wrong inter-project dependencies, as it is easily fixable in the future).
2. Thanks for testing it extensively and thanks for asking to double check it. We should do proper QA here if this changes so many files. @vitvakatu, do you think you could do QA here and check if this PR does not introduce any problems on your end when using the build script (run build / run watch)? You are really good at spotting issues in your QAs, so I thought I'd ask you for a small help here :) 
3. Changelog is needed for user-facing things, including technological changes (then users are considered users of our libraries). So visible changes to product, libraries, our rendering framework, should be in the changelog. This PR does not introduce any of such change, and thus, we should skip the changelog check here :) 
4. I believe the current solution is good enough. Watch script is the main tool for us to have fast incremental rebuilds, so if it builds tailwind once, no one really cares.

File: app/ide-desktop/lib/dashboard/src/tailwind.css
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/tailwind.css]: not a generated file, but rather manually created from tailwind's preflight. unfortunately there are no good solutions using purely tailwind. i can investigate whether this is possible using postcss + autoprefixer which is the normal solution

the reason this is required is that they all need the `#dashboard` prefix (*except* the one rule for `body`), otherwise the styles affect the IDE, causing things like #5963.

and they can't be left out either, otherwise the dashboard looks wrong, since e.g. border resets are not applied

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/tailwind.css]: (edit: `autoprefixer` is the wrong tool, and i believe `postcss-nested` (suggested by SO) *may* work but will cause tailwind to emit warnings. maybe i should just add a comment above this code explaining why it's needed?)

**Comment by somebody1234** [comment]
so it turns out i need to make some small changes because the merge added the plugin from the other branch

**Comment by somebody1234** [comment]
i've noticed that all the `bundle.ts` only bundle a single module, whereas `watch.ts` bundle everything that's required. not sure if `bundle.ts` are outdated or what

**Comment by somebody1234** [comment]
just a heads up, i haven't tested `content` after the merge, but it *looks* like everything from the old PR has been removed

**Comment by somebody1234** [comment]
edited PR description to mention that i switched from `ts-node` to `tsx` because i figured it'd be more appropriate. it seems like `ts-node` is still required as an optional dependency (this doesn't mean it can be removed, it means it won't error when it fails to install) because of some convoluted dependency chain:

```
ts-node@10.9.1 dev optional peer
node_modules/ts-node
  peerOptional ts-node@">=9.0.0" from postcss-load-config@3.1.4
  node_modules/postcss-load-config
    postcss-load-config@"^3.1.4" from tailwindcss@3.2.7
    node_modules/tailwindcss
      dev tailwindcss@"^3.2.7" from enso-dashboard@0.1.0
      lib/dashboard
        enso-dashboard@0.1.0
        node_modules/enso-dashboard
          workspace lib/dashboard from the root project
```

which is a bit unfortunate but i'm not sure there's anything we can do about this

**Comment by vitvakatu** [comment]
QA: @vitvakatu

**Comment by PabloBuchu** [comment]
@somebody1234 currently this doesn't work. Please rebase on top of this PR https://github.com/enso-org/enso/pull/6050 and ping me for more QAint

**Comment by somebody1234** [comment]
note sure if rebasing is a good idea since rebases imply force-pushes
i can merge though

**Comment by somebody1234** [comment]
@PabloBuchu cannot repro issues, `./run ide watch` works (before merge, at least), `npm run watch-dashboard` also works, testing `./run ide watch` again now and will test `./run ide build` soon

**Comment by somebody1234** [comment]
ok so `watch` seems to work, something weird seems to be happening on rebuilds... it *looks* like the css still works, but the IDE ui seems to be overwritten by the dashboard ui...

... on second thought, i think the build is overwriting content's `index.js` with dashboard's `index.js`. the simplest solution would be to rename dashboard's entrypoint but i'm not sure if that's an ideal solution

**Comment by somebody1234** [comment]
oh wait, `dashboard.tsx` isn't supposed to be included on desktop...

**Comment by somebody1234** [comment]
@PabloBuchu issues should be fixed now?

**Comment by vitvakatu** [comment]
I'll check in a few hours, was busy with other stuff.

**Comment by vitvakatu** [comment]
## QA Report 🔴 

The `./run ide watch` works fine, and reacts on changes nicely.

However, `./run ide build` does not finish running on my machine. (I left it for 30+ minutes, it didn't finish)

The last message I see is

```
INFO ide_ci::program::command: npmℹ️ Electron Builder is done. Result: [
 INFO ide_ci::program::command: npmℹ️   '/Users/ilyabogdanov/projects/enso/dist/ide/enso-mac-2023.1.1-dev.dmg'
 INFO ide_ci::program::command: npmℹ️ ]
```

This is not reproducible on the current develop, so I assume the issue is with this PR. I did a clean build after `./run git-clean` and removing `target` dir.

**Comment by somebody1234** [comment]
unfortunately i'm not quite sure what might be causing `./run ide build` to hang :/
i don't think i've changed any code related to electron-builder, and it's not hanging for me on linux

**Comment by somebody1234** [comment]
i can try `./run git-clean` + removing `target` if that may be the cause, but it would take quite a while

**Comment by somebody1234** [comment]
more notes:
- the esbuild scripts import from e.g. `../dashboard` and `../content` rather than importing through the modules themselves. i'm assuming this is acceptable as the modules shouldn't expose the esbuild scripts
- the esbuild scripts that include tailwind generation for `dashboard` are at the top level, rather than in the nested `authentication` module where they properly belong. about the organization of the `dashboard/src/authentication/` module, i'm not sure if this is acceptable for now, or if we should be restructuring it now since we're changing the build scripts here anyway

**Comment by somebody1234** [comment]
also re: testing, the following tests should probably also be performed:
- built binary with `-authentication` flag
  - ideally watch with `-authentication` flag, but i'm not sure if it's possible to pass flags to *only* the final executable in watch mode. if you do `-- -authentication`, some other CLI tool errors
- `npm run watch-dashboard` in the `app/ide-desktop/` directory (formerly `npm run dev`)

**Comment by somebody1234** [comment]
ah, to test authentication in `watch` mode you replace `"authentication": false` with `"authentication": true` in `app/ide-desktop/lib/content-config/src/config.json`

**Comment by vitvakatu** [comment]
@somebody1234 you can use --ide-option flag of ./run script to pass custom CLI arguments to the electron

**Comment by somebody1234** [comment]
more notes: because this switches from the old dedicated http server to esbuild's builtin http server, there may be some latency (~100ms?) every time a page is reloaded. i'm not sure whether that means we should be switching back to the old server? (this would probably depend on how often the page will be reloaded, if ever. afaict the dashboard uses client-side routing, which should *never* navigate away, and so the only time a page will be reloaded is for the live-server functionality used in the in-browser cloud frontend)

**Comment by somebody1234** [comment]
@vitvakatu ah TIL, thanks

**Comment by sylwiabr** [comment]
@somebody1234 @wdanilo what is the status of this? I understand this PR will allow me to finally test Cloud dashboard and authentication in Electron application? And we are not merging fix from here https://github.com/enso-org/enso/pull/6050 because there is an ugly copy-paste of code and this is solving the problem? When it will be merged then? We have to test the dashboard and authentication ASAP as some of the libraries features will rely on it. Since it is touching previous work of @mwu-tow maybe he should review it as well?

**Comment by somebody1234** [comment]
it should be ready but doesnt seem to be working for vitvakatu, so it probably still needs some QA

**Comment by somebody1234** [comment]
just completely cleaned and rebuilt and `./run ide build` is still successful for me (linux)
so it might be nice to get another macos user to confirm that `./run ide build` is indeed not broken?
(or if it is broken, i might need more information on how exactly it's failing, as i cannot reproduce the error)

**Comment by vitvakatu** [comment]
I spent some time debugging `./run build` on my machine. My findings:
1. The following command hangs: `cd "/Users/ilyabogdanov/projects/enso/app/ide-desktop" && ENSO_BUILD_GUI="/Users/ilyabogdanov/projects/enso/dist/gui" ENSO_BUILD_ICONS="/var/folders/kq/4hzyyjxd2lx24vckh9z6_l5c0000gn/T/.tmpEloMvG" ENSO_BUILD_IDE="/Users/ilyabogdanov/projects/enso/dist/ide" ENSO_BUILD_PROJECT_MANAGER="/Users/ilyabogdanov/projects/enso/dist/backend" PYTHON_PATH="/Users/ilyabogdanov/.pyenv/shims/python2" "/Users/ilyabogdanov/Library/Caches/fnm_multishells/67348_1679570731174/bin/npm" "--color" "always" "--yes" "--workspace" "enso" "run" "dist" "--" "--mac"`
2. `npm run dist` runs a script `enso/app/ide-desktop/lib/client/dist.ts`. This script works till the last line (I added a `console.log` at the end, and it works).
3. This is not the issue with our Rust code, as running the command in the console also hangs.
4. It is not caused by `npx`, though there are multiple similar [issues](https://github.com/electron-userland/electron-builder/issues/6411) in `electron-builder` repo. I've tried removing `npx` and running just `tsx dist.ts` – it still hangs.
5. It is not caused by npm. Running the same command directly (`npx tsx dist.ts -- --mac`) also hangs.
6. Finally, removing `const result = await electronBuilder.build(cliOpts)` from `enso/app/ide-desktop/lib/client/electron-builder-config.ts` solves the issue. I don't understand how it is possible.

Any thoughts?

**Comment by vitvakatu** [comment]
Using the most recent `electron-builder` (23.6.0) doesn't help.

**Comment by somebody1234** [comment]
worst case we could probably add `process.exit(0)` on the last line of `dist.ts`?
i'm assuming it's hanging because some other process isn't exiting as it should be... could you possibly search through running processes for anything that contains `node`?

**Comment by somebody1234** [comment]
alternatively you could try adding `console.log`s in `function build` in `lib/client/node_modules/app-builder-lib/out/index.js` to try and narrow down where it's hanging

**Comment by vitvakatu** [comment]
Here is a list of processes:
```
60427 ttys000    0:00.07 node /Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/.bin/tsx dist.ts --mac
60428 ttys000    0:07.23 /Users/ilyabogdanov/Library/Application Support/fnm/node-versions/v18.14.1/installation/bin/node --require /Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/tsx/dist/preflight.cjs --loader file:///Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/tsx/dist/loader.js dist.ts --mac
60429 ttys000    0:00.03 /Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/esbuild/lib/downloaded-@esbuild-darwin-arm64-esbuild --service=0.17.12 --ping
```

`build` function executes till the very end.

**Comment by vitvakatu** [comment]
For the reference, here is a list processes on develop:
```
66793 ttys000    0:00.17 node /Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/.bin/ts-node dist.ts --mac
66794 ttys000    0:45.89 node --require /Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/ts-node/dist/child/child-require.js --loader file:///Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/ts-node/child-loader.mjs /Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/ts-node/dist/child/child-entrypoint.js --brotli-base64-config=G9MRQByFcZvz1VZWpBwN5x3x7y31y2/jmxkpS9GYqauqV2/roagYSzLo8rpJ5IxXRw/NotPV16ple+/+rb2e6d0npayoGuIeWabgKGQ2IiXnMQpnV0iM5LWYy3hZWEh4bF39VL7pDYKt/JNAFNUBl1BcWYhUE9vWk1HjVTosxrTZvkvhN0gH67hLuQ01WVtzjZ5gMTRO88xaJbTNNLZW5bOgg2BH8HmIq8I5q8UlCHTtPurD+l8M0dGTdodTN0d2dZonOogsXbywaV8akPIHvN7fsWas1t/IXWLuKqlUYQSGmukd1/dGBxHClAqNLvCjPqBGS0VAH/rWZ8E7qZ5WKMGTUCWJk7iCI35r0Ya+Tf8MsdRBS5582dR+6x41M9qXEtycw+a5y4jT59WCJ9ESDowWj2etgmmpkWDZBns3RN71r4g1LJUEnPKpdpip3vIo5pIVTVKgTx8cNcfaplijgwY9mv+yM/8W18I7+HSYpln0mbtmX5lLb3OADPt3njSDwsX8JEFtuLJzHYVsegCKfL3UOS7zfWB0NYLGlF2nmAXvWvzaae1qGjoX1I9s5135ZqRGzAvXxdaYRbowYSfOvtPZKIIu2tQwnSSREzvK6it7xrIF8Ck/iOVjtVRuskRRYd4Iqu3WSuUoUvgU5L4wp9UGdielkXV3qanB5ZpuNbMQ56uU9FdJrVhiS83ICfLWR3yazuiF4IllL4XxA9mG+MpYruOjzrF2Nql7nHJbOa7XcaRdRHOgGwZoUNDVDoh4cyOgIqkFHu/EBQ5GrhEKA9cuDqKMjr0dBrV7WGngs8E3yox3HMB0lIdcsmfudV5ZHeVCXhb0Yl/xpCjITM3SA+IEh10EoJHsIKXtNCCJ1JZtERKzPQm9cFrJwCUuUMUAlm0fAThRiCLYUAmLAdpfmMuOr/4Xw5roGn5NgS+Gq1NOYti2ZrUkJUFjigU2OUI0I6KIFZ7gUGWSYF02GMfjdf6DIJblEdqoe1TOyF0qD5BwwZTs42YgC6tG62UTKtqqtVTL1qmidGNPmwOnszJXEiPHKC5UUqnkWtcf5G0rjRVFkWopMXppcS9W0ctLT2tea1RKGPccItJ7dc492EvS1XTaRZsJUyWYYnAPYmpUwX53eCU1dFMVCEG0oKNVv/QEI+oCn/M6tN61tI4YsWQaXAO1GBt6igTdJ9MtKLb1wFxJ438= dist.ts --mac
```
Adding `process.exit(0)` certainly helps.

**Comment by somebody1234** [comment]
hmm... looks like it might be `/Users/ilyabogdanov/projects/enso/app/ide-desktop/node_modules/esbuild/lib/downloaded-@esbuild-darwin-arm64-esbuild --service=0.17.12 --ping`?
could you run that command by itself and see what happens?
if that hangs, maybe it's an issue with newer versions of esbuild, in which case you could try installing older versions like `npm i esbuild@0.17.11` and see if any of `0.17.x` works
(remember to make sure the `npm i` actually installs the specified version (`cat node_modules/esbuild/package.json`))

**Comment by somebody1234** [comment]
... on second thought the issue might be a missing binary. @vitvakatu could you manually add `esbuild-darwin-arm64` to `optionalDependencies` in the `package.json` of each module with esbuild as a dependency and see if that fixes it? (which should be the modules with esbuild scripts, so `client`, `content`, and `dashboard`)

**Comment by vitvakatu** [comment]
> could you run that command by itself and see what happens?

It hangs.

> it's an issue with newer versions of esbuild, in which case you could try installing older versions

I've tried a few, still hangs.

> could you manually add esbuild-darwin-arm64 to optionalDependencies in the package.json of each module with esbuild as a dependency and see if that fixes it?

Doesn't fix it.

**Comment by somebody1234** [comment]
[this](https://github.com/vercel/turbo/issues/1769) says it might be to do with rosetta? so something might be pulling in the wrong version (x64 or arm64)

**Comment by somebody1234** [comment]
but also it turns out the naming convention for the binary packages was changed for 0.16+. @vitvakatu could you try again (maybe with a fresh `npm i`?)
not sure whether this change will do anything, since it didn't change anything for me (i got "up to date" when running `npm i`)

**Comment by vitvakatu** [comment]
@somebody1234 unfortunately, it still hangs. (after a clean build)

**Comment by somebody1234** [comment]
so according to [this](https://github.com/evanw/esbuild/issues/985) issue on esbuild, the long-running service seems to not be the problem?

but the node.js script *does* run until the end as well...

i think maybe `process.exit(0)` may be the best solution here then, if you've confirmed that everything still works fine after a clean build with `process.exit(0)` at the appropriate location

**Comment by vitvakatu** [comment]
Yes, adding `process.exit(0)` at the end of `dist.ts` fixes the issue. However, this is not a proper solution. We should create an issue and leave a comment there, explaining why this is needed.

**Comment by vitvakatu** [comment]
Also, according to the issue, can't we use `buildSync` instead?

**Comment by somebody1234** [comment]
actually, i think it's `npx tsx` calling esbuild (to transpile `dist.ts` and its imports)... so actually this might all be a red herring.
(i've looked into `dist.ts` and it seems like it doesn't)

but if it's not `esbuild` that's hanging up `node`, then... i'm not sure what else could be.

you might have luck running [why-is-node-running](https://github.com/mafintosh/why-is-node-running) if you can get that to work? probably by doing `npm i why-is-node-running`, importing it as `import log from 'why-is-node-running'` at the start of `dist.ts` and doing a `log()` at the end?

i have no idea what could be going wrong, because i just (re-)confirmed that `npx tsx dist.ts --linux` works for me on this branch (`5d83374c`) - maybe the best solution is just switching back to `ts-node`...

**Comment by somebody1234** [comment]
(the latest commit is just a minor, mostly stylistic, change, so it most likely won't fix anything)

**Comment by somebody1234** [comment]
the electron-builder issue you linked suggests using `npm exec` instead of `npx tsx`/`tsx` so that it knows
other options are doing `sudo strace -p <pid of the tsx dist.ts process>`, or `lsof | grep <pid of the tsx dist.ts process>` (or the other processes if the `tsx` one turns up nothing)

**Comment by somebody1234** [comment]
oh, or maybe change the `dist` script to `npx --node-options=--inspect tsx dist.ts` and open up a debugger (e.g. chrome://inspect/)

**Comment by PabloBuchu** [comment]
@somebody1234 @vitvakatu I've pushed the fix for now for hanging mac building. Can you confirm that is working. @wdanilo could you take a look and approve?

**Comment by somebody1234** [comment]
linux results:
- `enso.AppImage`: working
- `enso.AppImage -authentication`: working
- `npm run watch-dashboard`: still working
- `./run ide watch`: still working, css does get rebuilt on changes too
- `./run ide watch` with `"authentication": true`: still working, css does get rebuilt on changes too

(haven't tested actually logging in, i can do it if needed though)

**Comment by wdanilo** [comment]
> @somebody1234 @vitvakatu I've pushed the fix for now for hanging mac building. Can you confirm that is working. @wdanilo could you take a look and approve?

@vitvakatu is assigned as Q/A tester, so he should report that it is working first. Then, I'll do a fast second code review and we will be done.

**Comment by indiv0** [comment]
mac-arm64 results:
- `./run ide build`: works, no hang
- `enso.AppImage`: runs
- `npm run build-dashboard`: runs
- `npm run watch-dashboard`: runs
- `./run ide watch`: works, but I see no changes in `tailwind.css` when `Component` `className`s are changed

cc: @somebody1234 @PabloBuchu

**Comment by somebody1234** [comment]
@indiv0 just to confirm, "no changes" means that no changes happen *after* a `./run ide build` has been done after the edits have been made?

**Comment by indiv0** [comment]
@somebody1234 sorry, typo in my post. It's `./run ide watch` (not `build`) that doesn't seem to update the `tailwind.css`. If I:
- run `./run ide watch`,
- wait for it to build,
- make edits to the components (e.g., add a `text-3xl` to a `div`)
- wait for it to build,
- the `div` has the new `className` in the Electron window, but the `tailwind.css` doesn't seem to be updated accordingly

**Comment by somebody1234** [comment]
ah, right. that's intentional since the *proper* solution would be to make tailwind rebuild on every change to every `.tsx` file...
this shouldn't be an issue as tailwind builds are currently fairly fast, 
on the other hand it *will* be rebuilt every time *any* file is changed, *and* on every single reload.

on the other hand, having to stop and restart `watch-dashboard` on tailwind changes *has* been kinda painful, and so it'd be great if tailwind was re-generated every time, at least for `watch-dashboard`

@wdanilo i think you were the one that brought up build times in the previous issue, so... what are your thoughts on this?

**Review by vitvakatu — APPROVED** [review]
QA: passed

**Comment by PabloBuchu** [comment]
@wdanilo FYI QA is passed, all checks are green when you have time please take a look

File: app/ide-desktop/lib/client/watch.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/watch.ts]: what is "this" in the sentence "this is actually undefined"?

File: app/ide-desktop/lib/client/watch.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/watch.ts:84]: As we agreed, dashboard should not be a separate bundle landing in assets, right? Both cloud and desktop versions should just bundle `content` and dashboard should be dependency of content (which needs to probably be bundled before loading to content, but not to "assets", as `content` should be the main entry point always). We agreed to keep it for now this way to have these 2 entry points. However, when do we plan to unify it and use `content` as the main entry point and where is a task for it? We need to fix it, as otherwise we are growing bigger and bigger tech debt here, so I just want to be sure this is planned as soon as possible.

File: app/ide-desktop/lib/dashboard/bundle.ts
Comment (wdanilo) [inline app/ide-desktop/lib/dashboard/bundle.ts:15]: no sections

File: app/ide-desktop/lib/dashboard/package.json
Comment (wdanilo) [inline app/ide-desktop/lib/dashboard/package.json:11]: is the type checking and eslint integrated to the build script (and thus CI)?

**Review by wdanilo — APPROVED** [review]
Just to double check:
1. The comment I left in the code is important. We still bundle dashboard to "assets" and we need to stop using it as separate target asap. We can merge this PR, but the proper fix needs to be planned.
2. I understand that we are 100% sure that the tailwind generation is done once when running watch and watch is not affected by any additional actions (in comparison to the state before this PR), right? :)

File: app/ide-desktop/lib/client/watch.ts
Comment (PabloBuchu) [reply app/ide-desktop/lib/client/watch.ts:84]: As we are not planning to open `https://cloud.enso.org` the separate entrypoint is mainly being used for development and review/debug stuff. 

We are planning to completely remove it as soon as we have pending dashboard PR merged.

Comment (PabloBuchu) [reply app/ide-desktop/lib/client/watch.ts:84]: Here is the task for it https://github.com/enso-org/cloud-v2/issues/345

Comment (wdanilo) [reply app/ide-desktop/lib/client/watch.ts:84]: We are not planning to provide users with access to `https://cloud.enso.org` for now, but it still should work. Hopefully sooner than later we will be able to work in the cloud. Anyway, I understand that the plan is that `content` will be used as the main entry point for both electron and the web version right?

Anyway, this build script puts the results into "assets", so it does not look like it's for dev/testing only, as "assets" is bundled with the final app.

Comment (wdanilo) [reply app/ide-desktop/lib/client/watch.ts:84]: Anyway, I understand that after the PR you mentioned is merged, we will move it this way that content will be the main entry point for both electron and the web version, is that correct?

Comment (PabloBuchu) [reply app/ide-desktop/lib/client/watch.ts:84]: yes, we need to push dashboard to be ready then we can polish and fix this technology debt

File: app/ide-desktop/lib/dashboard/package.json
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/package.json:11]: i don't think so, we have a `prettier` check in the lint CI action but nothing else:

https://github.com/enso-org/enso/blob/6b9cbeacb2caa7606d91fda90ec3ec40acb39c70/build/cli/src/lib.rs#L835

we seem to be manually creating a wrapper module for prettier [here](https://github.com/enso-org/enso/blob/1a569223aaf13e8c3ca40cd03df3afa227fea625/build/build/src/prettier.rs) so i'd assume someone would have to make similar wrappers for `tsc` and `eslint`

**Comment by somebody1234** [comment]
@wdanilo `watch` *is* affected by changes to `tailwind.css`, *however* it should only happen when `tailwind.css` or `tailwind.config.js` are updated - so if they *aren't* changed and the current code still rebuilds it, it is a bug

**Comment by PabloBuchu** [comment]
@indiv0 @somebody1234 this PR a bug from yesterday (reverted here https://github.com/enso-org/enso/pull/6127). THe bug wasn't present before. Can you both try and reproduce this on your machines? Also @vitvakatu can you once more try to build and run enso from this branch?

Problems occurs when you have `-authentication=true` and try to run from built /dist

Screenshot 2023-03-30 at 12 07 34
Screenshot 2023-03-30 at 12 07 26
Screenshot 2023-03-30 at 12 07 14

**Review by PabloBuchu — CHANGES_REQUESTED** [review]
see the other comment

**Comment by indiv0** [comment]
Looks like the `"root"` of the application is getting loaded twice? This is probably the cause. I'll try to see if I can figure out why.

Screenshot 2023-03-30 at 16 02 28

File: app/ide-desktop/lib/dashboard/src/authentication/src/index.tsx
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/src/authentication/src/index.tsx]: @somebody1234 Commenting out these lines seems to make the error non-reproducible. I think they may be responsible.

At a glance, it looks like we never reset the `ide.style.display` back to `""`. So it remains `"none"` after authentication is complete and we do `appInstance.run()`. So EnsoGL tries to render to a non-visible element, which I'm assuming is not possible.

Making the IDE visible prior to calling `appInstance.run()` seems to fix it:
```diff
diff --git a/app/ide-desktop/lib/content/src/index.ts b/app/ide-desktop/lib/content/src/index.ts
index 3fc612a43..34d3db0e7 100644
--- a/app/ide-desktop/lib/content/src/index.ts
+++ b/app/ide-desktop/lib/content/src/index.ts
@@ -143,7 +143,9 @@ class Main {
                 ) {
                     const hideAuth = () => {
                         const auth = document.getElementById('dashboard')
+                        const ide = document.getElementById('root')
                         if (auth) auth.style.display = 'none'
+                        if (ide) ide.style.display = ''
                     }
                     /** This package is an Electron desktop app (i.e., not in the Cloud), so
                      * we're running on the desktop. */
```

**Comment by indiv0** [comment]
> Looks like the `"root"` of the application is getting loaded twice? This is probably the cause. I'll try to see if I can figure out why.
> 
> Screenshot 2023-03-30 at 16 02 28

Worth nothing that we are rendering the scene twice over for some reason, though this isn't the cause of the bug.

**Review by PabloBuchu — APPROVED** [review]

**Comment by wdanilo** [comment]
Double `root` is bad and we should fix it.

**Comment by indiv0** [comment]
> Double `root` is bad and we should fix it.

@wdanilo Double root is fixed with the workaround discussed in https://github.com/enso-org/enso/pull/6132 and the proper fix is being tracked here https://github.com/enso-org/cloud-v2/issues/366

It's not an issue specific to this PR

**Comment by PabloBuchu** [comment]
QA 🟢  I will merge then as soon as the checks finishes

**Pull request merged**
